24F-AI-125 Mariyam

Open-ended lab

Project 02: Health Risk Classification

In [1]:
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

In [2]:
data = pd.read_csv("diabetes.csv")
X = data.drop(columns="Outcome")
y = data["Outcome"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [ ]:
#SVM
svm_model = make_pipeline(
    SimpleImputer(strategy="median"),
    StandardScaler(),
    SVC()
)

svm_model.fit(X_train, y_train)
svm_train_accuracy = accuracy_score(y_train, svm_model.predict(X_train))
svm_test_accuracy = accuracy_score(y_test, svm_model.predict(X_test))

print("SVM Training Accuracy:", round(svm_train_accuracy, 3))
print("SVM Test Accuracy:", round(svm_test_accuracy, 3))

SVM Training Accuracy: 0.896
SVM Test Accuracy: 0.838


In [ ]:
#Decision Tree
dt_model = make_pipeline(
    SimpleImputer(strategy="median"),
    DecisionTreeClassifier(random_state=42)
)

dt_model.fit(X_train, y_train)
dt_train_accuracy = accuracy_score(y_train, dt_model.predict(X_train))
dt_test_accuracy = accuracy_score(y_test, dt_model.predict(X_test))

print("Decision Tree Training Accuracy:", round(dt_train_accuracy, 3))
print("Decision Tree Test Accuracy:", round(dt_test_accuracy, 3))

Decision Tree Training Accuracy: 1.0
Decision Tree Test Accuracy: 0.812


In [ ]:
#Comparison
svm_gap = svm_train_accuracy - svm_test_accuracy
dt_gap = dt_train_accuracy - dt_test_accuracy

print("SVM")
print("train accuracy:", round(svm_train_accuracy, 3))
print("test accuracy:", round(svm_test_accuracy, 3))
print("gap:", round(svm_gap, 3))

print()

print("Decision Tree")
print("train accuracy:", round(dt_train_accuracy, 3))
print("test accuracy:", round(dt_test_accuracy, 3))
print("gap:", round(dt_gap, 3))

SVM
train accuracy: 0.896
test accuracy: 0.838
gap: 0.058

Decision Tree
train accuracy: 1.0
test accuracy: 0.812
gap: 0.188


decsision tree scored 1.0 accuracy upon training means 100% memozied but on testing phase it dropped to 0.812approx 81% means model is performing well in traing and huge gap in testing. which means it overfits.
whereas, svm in kind of giving generalized performance as its accuracy scores are optimized.
hence SVM is better for this very dataset.